### ISSUE TITLE
get and set params in a sklearn compatible method

### ISSUE THREAD

**Issue by rpreen** [rpreen/xcsf#76]
Provide `get_params()`, `set_params()` and modify the constructor on the Python library to allow setting parameters in a [sklearn compatible](https://scikit-learn.org/stable/developers/develop.html#rolling-your-own-estimator) way.

**Comment by dpaetzel** [comment]
Hi! Thank you for working on this! Are you planning to modify the constructor in the bindings themselves or do you intend to use a Python wrapper around the XCS object? I wrote a wrapper a few weeks ago (although not a particularly *nice* wrapper, just so that I barely implement the interface and only for regression tasks so far) and could share my code, maybe you can spare yourself from writing some boilerplate like that?

**Comment by dpaetzel** [comment]
> whether the constructor should take `kwargs`

I think if you want to implement the sklearn interface fully then you won't get around having a massive long list of arguments … In the [sklearn docs](https://scikit-learn.org/stable/developers/develop.html#instantiation) they say

> In addition, every keyword argument accepted by __init__ should correspond to an attribute on the instance. Scikit-learn relies on this to find the relevant attributes to set on an estimator when doing model selection.

I guess if I had to do this myself, I'd script the boilerplate generation somehow (probably using Python and a regex to extract all the parameter names). This way, if another parameter is added sometime, I could simply regenerate the boilerplate.

**Comment by rpreen** [comment]
Yeah, they say that, but actually they use a form of duck typing to check and if it passes their checking tool it's ok:

```python
from sklearn.utils.estimator_checks import check_estimator                                                                                                                         
import xcsf
xcs = xcsf.XCS(x_dim=1, y_dim=1, n_actions=1)
check_estimator(xcs)
```

Of course it fails this at the moment, but I was able to get past the parameter checks by using kwargs...

**Comment by rpreen** [comment]
Turns out there is no easy shortcut to passing the `check_estimator()` by using the JSON setters/getters because it compares the memory addresses of parameters so even though the values are identical it fails - why this is necessary for comparing ints and floats etc. I don't know.

https://github.com/scikit-learn/scikit-learn/blob/fe26a08c5c0902db80667bed2d72a506826417ab/sklearn/base.py#L121

**Comment by rpreen** [comment]
@dpaetzel I managed to get as far as getting it to work with sklearn's cross validation. What are the other functions that you were using that needed the wrapper?

**Comment by dpaetzel** [comment]
Awesome! :tada: 

Right now I'm using sklearn's [pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline).

```
from sklearn.pipeline import make_pipeline
from sklearn.compose import TransformedTargetRegressor
from sklearn.preprocessing import MinMaxScaler, StandardScaler

model = make_pipeline(
    MinMaxScaler(feature_range=(-1.0, 1.0)),
    TransformedTargetRegressor(regressor=XCSF(),
                               transformer=StandardScaler()))
model.fit(X, y)
```

I then give that pipeline (without fitting it, of course) to [Optuna's OptunaSearchCV](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.OptunaSearchCV.html).

In the past, I've used sklearn's [grid search](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV); the Optuna interface tries to be close to the grid search interface if I recall correctly.

**Comment by rpreen** [comment]
I have a [notebook](https://github.com/rpreen/xcsf/blob/get_set_params/python/notebooks/example_regression.ipynb) with an early stopping callback working.

And I have a [notebook](https://github.com/rpreen/xcsf/blob/get_set_params/python/notebooks/example_sklearn.ipynb) with XCSF working with sklearn pipelines and grid search.

BUT: apparently sklearn pipelines will not allow passing any parameters to `fit()` except `X_train` and `y_train`, so the callbacks can't be used there. I guess the only way around this is to not use callbacks, but instead have them as just more parameters in the constructor.

I haven't really used the sklearn functions such as pipelines etc. - is there any thing else that it would be useful to make sure works?
